Sources/Shared/MultipleImageFetcher.swift

```diff
@@ -0,0 +1,49 @@
+import Foundation
+
+// Fetch multiple images. This is ideal for prefetching.
+public class MultipleImageFetcher {
+
+  private var fetchers: [ImageFetcher] = []
+  private let fetcherMaker: () -> ImageFetcher
+
+  init(fetcherMaker: @escaping () -> ImageFetcher) {
+    self.fetcherMaker = fetcherMaker
+  }
+
+  /// Fetch multiple urls at once
+  ///
+  /// - Parameters:
+  ///   - urls: The urls to fetch
+  ///   - each: Called after each fetcher completed
+  ///   - completion: Called after all fetchers completed
+  public func fetch(urls: [URL],
+                    each: ((Result) -> Void)? = nil,
+                    completion: (([Result]) -> Void)? = nil) {
+    self.fetchers = urls.map { _ in
+      return self.fetcherMaker()
+    }
+
+    var results: [Result] = []
+
+    zip(fetchers, urls).forEach { fetcher, url in
```

Comment:
@onmyway133 You couldn't do it without functional programming, didn't you? 😄

Reply:
@vadymmarkov 😛